File: c++/121_best_time_to_buy_and_sell_stock.cpp
#include <gtest/gtest.h>
#include <iostream>
#include <limits>
#include <vector>

class Solution {
   public:
    int maxProfit(const std::vector<int>& prices) {
        int minPrice  = std::numeric_limits<int>::max();
        int maxProfit = 0;
        for (auto price : prices) {
            minPrice  = std::min(minPrice, price);
            maxProfit = std::max(maxProfit, price - minPrice);
        }
        return maxProfit;
    }
};

int main(int argc, char* argv[]) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

TEST(test, case1) {
    Solution solution;
    EXPECT_EQ(solution.maxProfit({7, 1, 5, 3, 6, 4}), 5);
}

File: python/121_best_time_to_buy_and_sell_stock.py
#!/usr/bin/env python3

import sys


class Solution:
    def maxProfit(self, prices):
        minPrice = sys.maxsize
        maxProfit = 0
        for price in prices:
            minPrice = min(minPrice, price)
            maxProfit = max(maxProfit, price - minPrice)
        return maxProfit


if __name__ == '__main__':
    print(Solution().maxProfit([7, 1, 5, 3, 6, 4]))
